### PR TITLE
Fix broken browser detection on Linux when 2 browsers have same version #

### LINKED ISSUE #1400: Cypress does not find chromium-browser on linux
Environment: Ubuntu 16.04, Cypress 2.1.0

Repro steps: 

1. `sudo apt-get update && apt-get install chromium-browser`
2. `which chromium browser` (in my case this output `/usr/local/bin/chromium-browser`)
3. `./node_modules/.bin/cypress run --browser chromium-browser`

```
Results: Browser: 'chromium-browser' was not found on your system.
```

Only electron and chrome are being found. The same is true when running cypress open and seeing which browsers are available in the dropdown.

### PR DESCRIPTION
**Pull request by flotwig**
Fixes #1400

image

### CODE REVIEW

**Comment by bahmutov** [comment]
oh my, if this is true, great catch!

File: packages/launcher/lib/detect.ts [hunk 1]
```diff
@@ -73,12 +73,13 @@ function checkOneBrowser(browser: Browser) {
...
 /** returns list of detected browsers */
-function detectBrowsers(): Bluebird<Browser[]> {
+function detectBrowsers(goalBrowsers?: Browser[]): Bluebird<Browser[]> {
   // we can detect same browser under different aliases
-  // tell them apart by the full version property
+  // tell them apart by the name and the version property
   // @ts-ignore
-  const removeDuplicates = uniqBy(prop('version'))
-  return Bluebird.mapSeries(browsers, checkOneBrowser)
+  if (!goalBrowsers) goalBrowsers = browsers
```
Comment (bahmutov) [inline packages/launcher/lib/detect.ts]: hmm, we don't require `{ }` around code blocks? like it should be
```
if (!goalBrowsers) {
  goalBrowsers = browsers
}
```

Comment (flotwig) [reply packages/launcher/lib/detect.ts]: Looks like the linter doesn't run on specfiles, I'll see about changing that

File: packages/launcher/test/unit/detect_spec.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 const detect = require('../../lib/detect').default
-const log = require('debug')('cypress:launcher:test')
 const os = require('os')
+import { log } from '../log'
```
[the remark this answers is not in the document]
Comment (flotwig) [reply packages/launcher/test/unit/detect_spec.ts]: it is used, line 13 and 17

Comment (bahmutov) [inline packages/launcher/test/unit/detect_spec.ts]: ohh, my bad, yes you have replaced `const log = ` with `import { log }` which makes sense.

File: packages/launcher/test/unit/linux/spec.ts [hunk 3]
```diff
@@ -30,4 +57,31 @@ describe('linux browser detection', () => {
...
+      {
+        name: 'foo-browser',
+        version: '100.1.2.3',
+        path: 'foo-browser',
+        majorVersion: '100'
+      }
+    ]
+    
+    detect(goalBrowsers).then(browsers => {
+      log('Browsers:', browsers)
```
[the remark this answers is not in the document]
Comment (flotwig) [reply packages/launcher/test/unit/linux/spec.ts]: like this?

```ts
log('Browsers:', browsers, 'Expected browsers:', expected)
```

[the remark this answers is not in the document]
Comment (bahmutov) [reply packages/launcher/test/unit/linux/spec.ts]: more like
```
log('Browsers: %o', browsers)
log('Expected browsers: %o', expected)
```

File: packages/launcher/lib/detect.ts [hunk 4]
```diff
@@ -73,12 +73,15 @@ function checkOneBrowser(browser: Browser) {
 }
 
 /** returns list of detected browsers */
-function detectBrowsers(): Bluebird<Browser[]> {
+function detectBrowsers(goalBrowsers?: Browser[]): Bluebird<Browser[]> {
   // we can detect same browser under different aliases
-  // tell them apart by the full version property
+  // tell them apart by the name and the version property
   // @ts-ignore
```
[the remark this answers is not in the document]
Comment (flotwig) [reply packages/launcher/lib/detect.ts]: not sure, it was there originally. it looks like `props` or `prop` throw this TS warning:

```
[ts]
Argument of type '<T>(obj: Record<"name", T>) => T' is not assignable to parameter of type '(a: {}) => {}'.
  Types of parameters 'obj' and 'a' are incompatible.
    Property 'name' is missing in type '{}' but required in type 'Record<"name", {}>'. [2345]
```

It seems like the ramda authors are aware, they have some comments in their .d.ts file about typescript versions

File: packages/launcher/test/unit/linux/spec.ts [hunk 5]
```diff
@@ -30,4 +57,31 @@ describe('linux browser detection', () => {
...
+      log('Expected browsers:', expected)
+      expect(browsers).to.have.deep.members(expected)
+      done()
```
[the remark this answers is not in the document]
Comment (brian-mann) [reply packages/launcher/test/unit/linux/spec.ts]: Wanted to chime in here definitely agree - the one exception is where you want to test the FAILURE state of an async event not connected to the promise. Can show examples if you're curious next time we pair.

**Comment by bahmutov** [comment]
Also can do “expect(promise).to.eventually deep.equal(...)” syntax

**Pull request merged**